core/io/logger.cpp

```diff
@@ -266,3 +267,354 @@ CompositeLogger::~CompositeLogger() {
+	// Anything that gets delivered while we're writing these messages gets to wait for next frame.
+	// (We could loop until the buffer is empty)
+	Vector<Dictionary> buffered_logs_mirror;
+	{
+		MutexLock lock(mutex);
+		SWAP(buffered_logs_mirror, buffered_logs);
+	}
+
+	// Dispatch to all the buffered callables, in a threadsafe manner
+	// Any buffered callables that disable themselves stop getting messages ASAP
+	// any buffered callables that get attached might start getting messages midway through
+	// we're ok with that, it's still a chronologically coherent block
+	for (int i = 0; i < buffered_logs_mirror.size(); ++i) {
+		// This is the same index-as-iterator-to-avoid-lock-issues dance we do in register_log_capture_nonthreadsafe()
+		int indexToSendTo = 0;
+
+		while (true) {
+			Callable callable;
+			{
+				MutexLock lock(mutex);
+				if (indexToSendTo >= captures_buffered.size()) {
+					// we done!
+					break;
+				}
+
+				callable = captures_buffered[indexToSendTo++];
+			}
+
+			dispatch_message(buffered_logs_mirror[i], callable);
+		}
+	}
+
+	if (prebuffering) {
+		// Buffering is done!
+		// Read the comment near the bottom of recalculate_state for a possible race condition and how it's dealt with.
+		// IT IS REALLY IMPORTANT THAT THIS HAPPENS ONLY ON THE MAIN THREAD.
+		MutexLock lock(mutex);
+		prebuffering = false;
+		recalculate_state();
+	}
+}
+
+void UserLogManagerLogger::process(const Dictionary &p_message) {
+	Vector<Callable> captures_nonthreadsafe_mirror;
+	{
+		// Shove another item into the buffer, if we need to.
+
+		// We also atomically grab `captures_nonthreadsafe` to avoid race conditions, in the extraordinarily unlikely case where:
+		// * this function buffers a message, releases the lock, and is immediately preempted
+		// * in another thread, an already-running register_log_capture_nonthreadsafe dispatches our new message from the buffer
+		// * that same register_log_capture_nonthreadsafe gets to the end of the buffer and adds itself to captures_nonthreadsafe
+		// * this function resumes, reaches the captures_nonthreadsafe section, and sends the same message again
+
+		// register_log_capture_nonthreadsafe intentionally finishes the buffered log message replay atomically with adding a new handler to the vector
+		// so we do the same thing here, adding a message to the log atomically with copying the handlers
+		// Vector COW behavior prevents this from being expensive in the common case
+		MutexLock lock(mutex);
+		if (state == STATE_BUFFERING) {
+			buffered_logs.append(p_message);
+		}
+		captures_nonthreadsafe_mirror = captures_nonthreadsafe;
+	}
+
+	// Dispatch to all the nonthreadsafe callables at the moment we added to the buffer
+	// We actually don't have to care about cutesy threadsafety for once because we're working off a local copy of the captures list
+
+	for (int i = 0; i < captures_nonthreadsafe_mirror.size(); ++i) {
```

Comment:
You may be able to use a for range loop here (`for (... : ...)`). Also, we tend to prefer post-incrementation to pre-incrementation when we can't use `for` range loops.

Reply:
Done!